pkg/threescale/connectors/backend/cached_backend.go

```diff
@@ -0,0 +1,261 @@
+package backend
+
+import (
+	"fmt"
+	"net/http"
+	"sync"
+
+	backend "github.com/3scale/3scale-go-client/client"
+	"github.com/3scale/3scale-istio-adapter/pkg/threescale/metrics"
+	cmap "github.com/orcaman/concurrent-map"
+)
+
+const (
+	cacheKeySeparator = "_"
+	cacheKey          = "%s" + cacheKeySeparator + "%s"
+)
+
+// Cacheable - defines the required behaviour of a Backend cache
+type Cacheable interface {
+	Get(cacheKey string) (CacheValue, bool)
+	Set(cacheKey string, cv CacheValue)
+}
+
+// Limit captures the current state of the rate limit for a particular time period
+type Limit struct {
+	current int
+	max     int
+	// unix timestamp for end period
+	periodEnds int64
+}
+
+// CacheValue which should be stored in cache implementation
+type CacheValue map[string]map[backend.LimitPeriod]*Limit
+
+// CachedBackend provides a pluggable cache enabled 'Backend' implementation
+// Supports reporting metrics for non-cached responses.
+type CachedBackend struct {
+	ReportFn      metrics.ReportMetricsFn
+	cache         Cacheable
+	mutex         sync.RWMutex
+	hierarchyTree hierarchyTree
+}
+
+type LocalCache struct {
+	ds cmap.ConcurrentMap
+}
+
+// map to represent parent --> children metric relationship
+type metricParentToChildren map[string][]string
+
+// hierarchyTree - represents the relationship between service metrics and their children
+type hierarchyTree map[string]metricParentToChildren
+
+// NewCachedBackend returns a pointer to a CachedBackend with a default LocalCache if no custom implementation has been provided
+func NewCachedBackend(cache Cacheable, rfn metrics.ReportMetricsFn) *CachedBackend {
+	if cache == nil {
+		cache = NewLocalCache()
+	}
+	return &CachedBackend{
+		ReportFn:      rfn,
+		cache:         cache,
+		hierarchyTree: make(hierarchyTree),
+	}
+}
+
+// NewLocalCache returns a pointer to a LocalCache with an initialised empty data structure
+func NewLocalCache() *LocalCache {
+	return &LocalCache{ds: cmap.New()}
+}
+
+// Get entries for LocalCache
+func (l LocalCache) Get(cacheKey string) (CacheValue, bool) {
+	var cv CacheValue
+	v, ok := l.ds.Get(cacheKey)
+	if !ok {
+		return cv, ok
+	}
+
+	cv = v.(CacheValue)
+	return cv, ok
+}
+
+// Set entries for LocalCache
+func (l LocalCache) Set(cacheKey string, cv CacheValue) {
+	l.ds.Set(cacheKey, cv)
+}
+
+// AuthRep provides a combination of authorizing a request and reporting metrics to 3scale
+func (cb CachedBackend) AuthRep(req AuthRepRequest, c *backend.ThreeScaleClient) (Response, error) {
```

Comment:
Right, I figured that was the case anyway but dismissed the fact until we have the logic correct here

Reply:
I'd say better to be aggressive about that, because it makes understanding the logic easier, and explicits dependencies as parameters and results as returned values (unless some of those are stored in a `self` struct, but that also becomes apparent).